cfgmgr/buffermgrd.cpp

```diff
@@ -121,6 +135,14 @@ int main(int argc, char **argv)
 
     SWSS_LOG_NOTICE("--- Starting buffermgrd ---");
 
+#if defined(ASAN_ENABLED)
+    if (signal(SIGTERM, sigterm_handler) == SIG_ERR)
```

Comment:
>signal [](http://example.com/codeflow?start=8&length=6)

This code lines are repeated in each main(). Could you define a common main function, and calling into old main(), which renamed to like buffermgrd_main().

Reply:
Could you also elaborate on adding a signal handler for calling `__lsan_do_leak_check` explicitly? Leak checks are atuomatically done when the process is exiting gracefully, so this shouldn't be needed at all.

Reply:
> Could you also elaborate on adding a signal handler for calling `__lsan_do_leak_check` explicitly? Leak checks are atuomatically done when the process is exiting gracefully, so this shouldn't be needed at all.

There are a couple of problems here why it is not working as expected. All applications are running as daemons and handing requests in the loop. To exit gracefully they need to receive a request to shut down, exit from the loop, and deallocate all the memory previously allocated. Then exit from the main function and after that address sanitizer will trigger the leak checker. The main problem here is that most if not all applications are running an infinite loop. They are not exiting correctly from the loop and are simply killed with the SIGTERM signal. Because of that when address sanitizer starts leaks checking it reports all the memory that was allocated during the initialization of the application as a leaked. So reports had dozens of false-positive memory leaks.
Even if we had decided to change the applications and exit from the loop correctly on a signal this wouldn't solve the problem. Because most of the objects and raw memory allocated on the initialization are not deallocated. And we again would have dozens of leaks in the reports.
Some time ago I created a pull request that does described changes: https://github.com/Azure/sonic-swss/pull/2123. And then decided to close it and change implementation to call __lsan_do_leak_check from the signal handler directly. This allows triggering leaks checker without the need to exit the loop and change of application. It reports only memory that is definitely lost and doesn't care about the memory that still has valid pointers.